#Task 1: Preprocessing


In [1]:
#importing file to colab

#Given the text file (grail.txt) from the Web Text Corpus of NLTK

from google.colab import files 
uploaded = files.upload()

Saving grail.txt to grail.txt


In [2]:
#install nltk to colab

!pip install nltk

In [26]:
#library to do -numeric to string

!pip install inflect

In [56]:
#import and download necessary libraries

import nltk
nltk.download('punkt')
nltk.download('stopwords')

from nltk.tokenize import word_tokenize
from nltk.tokenize import sent_tokenize
import string
import inflect
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer

import re


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [4]:
#Reading the given file

with open("grail.txt") as f:
  original_text = f.read()

print(original_text)

SCENE 1: [wind] [clop clop clop] 
KING ARTHUR: Whoa there!  [clop clop clop] 
SOLDIER #1: Halt!  Who goes there?
ARTHUR: It is I, Arthur, son of Uther Pendragon, from the castle of Camelot.  King of the Britons, defeator of the Saxons, sovereign of all England!
SOLDIER #1: Pull the other one!
ARTHUR: I am, ...  and this is my trusty servant Patsy.  We have ridden the length and breadth of the land in search of knights who will join me in my court at Camelot.  I must speak with your lord and master.
SOLDIER #1: What?  Ridden on a horse?
ARTHUR: Yes!
SOLDIER #1: You're using coconuts!
ARTHUR: What?
SOLDIER #1: You've got two empty halves of coconut and you're bangin' 'em together.
ARTHUR: So?  We have ridden since the snows of winter covered this land, through the kingdom of Mercea, through--
SOLDIER #1: Where'd you get the coconuts?
ARTHUR: We found them.
SOLDIER #1: Found them?  In Mercea?  The coconut's tropical!
ARTHUR: What do you mean?
SOLDIER #1: Well, this is a temperate zone.
AR

1. Report the number of sentences and tokens contained in the file given as input.

In [66]:
#counting number of sentences

print(sent_tokenize(original_text))
print("Number of Sentences= ", len(sent_tokenize(original_text)))

['SCENE 1: [wind] [clop clop clop] \nKING ARTHUR: Whoa there!', '[clop clop clop] \nSOLDIER #1: Halt!', 'Who goes there?', 'ARTHUR: It is I, Arthur, son of Uther Pendragon, from the castle of Camelot.', 'King of the Britons, defeator of the Saxons, sovereign of all England!', 'SOLDIER #1: Pull the other one!', 'ARTHUR: I am, ...  and this is my trusty servant Patsy.', 'We have ridden the length and breadth of the land in search of knights who will join me in my court at Camelot.', 'I must speak with your lord and master.', 'SOLDIER #1: What?', 'Ridden on a horse?', 'ARTHUR: Yes!', "SOLDIER #1: You're using coconuts!", 'ARTHUR: What?', "SOLDIER #1: You've got two empty halves of coconut and you're bangin' 'em together.", 'ARTHUR: So?', "We have ridden since the snows of winter covered this land, through the kingdom of Mercea, through--\nSOLDIER #1: Where'd you get the coconuts?", 'ARTHUR: We found them.', 'SOLDIER #1: Found them?', 'In Mercea?', "The coconut's tropical!", 'ARTHUR: What 

In [68]:
#tokenization #word

original_text_token = word_tokenize(original_text)
print("Tokens in Original sample text are- ", original_text_token)
print("Number of tokens in original text =", len(original_text_token))

Tokens in Original sample text are-  ['SCENE', '1', ':', '[', 'wind', ']', '[', 'clop', 'clop', 'clop', ']', 'KING', 'ARTHUR', ':', 'Whoa', 'there', '!', '[', 'clop', 'clop', 'clop', ']', 'SOLDIER', '#', '1', ':', 'Halt', '!', 'Who', 'goes', 'there', '?', 'ARTHUR', ':', 'It', 'is', 'I', ',', 'Arthur', ',', 'son', 'of', 'Uther', 'Pendragon', ',', 'from', 'the', 'castle', 'of', 'Camelot', '.', 'King', 'of', 'the', 'Britons', ',', 'defeator', 'of', 'the', 'Saxons', ',', 'sovereign', 'of', 'all', 'England', '!', 'SOLDIER', '#', '1', ':', 'Pull', 'the', 'other', 'one', '!', 'ARTHUR', ':', 'I', 'am', ',', '...', 'and', 'this', 'is', 'my', 'trusty', 'servant', 'Patsy', '.', 'We', 'have', 'ridden', 'the', 'length', 'and', 'breadth', 'of', 'the', 'land', 'in', 'search', 'of', 'knights', 'who', 'will', 'join', 'me', 'in', 'my', 'court', 'at', 'Camelot', '.', 'I', 'must', 'speak', 'with', 'your', 'lord', 'and', 'master', '.', 'SOLDIER', '#', '1', ':', 'What', '?', 'Ridden', 'on', 'a', 'horse', '?

Answer 1-

Number of Sentences in sample file=  1881

Number of tokens in original text(sample file) = 16450


2. Convert the whole text to lower case and report the number of unique tokens present before and after lower casing in the input file.

In [71]:
#remove punctuations
#puncs = string.punctuation
#print(puncs)

#mapping = str.maketrans("","",puncs)
#original_text = original_text.translate(mapping)

beforelowering_text_token = word_tokenize(original_text)

print(len(beforelowering_text_token))

16450


In [76]:
#all to lower case
lowercase_text = original_text.lower()
print(lowercase_text)

scene 1: [wind] [clop clop clop] 
king arthur: whoa there!  [clop clop clop] 
soldier #1: halt!  who goes there?
arthur: it is i, arthur, son of uther pendragon, from the castle of camelot.  king of the britons, defeator of the saxons, sovereign of all england!
soldier #1: pull the other one!
arthur: i am, ...  and this is my trusty servant patsy.  we have ridden the length and breadth of the land in search of knights who will join me in my court at camelot.  i must speak with your lord and master.
soldier #1: what?  ridden on a horse?
arthur: yes!
soldier #1: you're using coconuts!
arthur: what?
soldier #1: you've got two empty halves of coconut and you're bangin' 'em together.
arthur: so?  we have ridden since the snows of winter covered this land, through the kingdom of mercea, through--
soldier #1: where'd you get the coconuts?
arthur: we found them.
soldier #1: found them?  in mercea?  the coconut's tropical!
arthur: what do you mean?
soldier #1: well, this is a temperate zone.
ar

In [77]:
#tokenization #word #unique tokens

lowercase_text_token = word_tokenize(lowercase_text)
print(lowercase_text_token)
print(len(lowercase_text_token))

['scene', '1', ':', '[', 'wind', ']', '[', 'clop', 'clop', 'clop', ']', 'king', 'arthur', ':', 'whoa', 'there', '!', '[', 'clop', 'clop', 'clop', ']', 'soldier', '#', '1', ':', 'halt', '!', 'who', 'goes', 'there', '?', 'arthur', ':', 'it', 'is', 'i', ',', 'arthur', ',', 'son', 'of', 'uther', 'pendragon', ',', 'from', 'the', 'castle', 'of', 'camelot', '.', 'king', 'of', 'the', 'britons', ',', 'defeator', 'of', 'the', 'saxons', ',', 'sovereign', 'of', 'all', 'england', '!', 'soldier', '#', '1', ':', 'pull', 'the', 'other', 'one', '!', 'arthur', ':', 'i', 'am', ',', '...', 'and', 'this', 'is', 'my', 'trusty', 'servant', 'patsy', '.', 'we', 'have', 'ridden', 'the', 'length', 'and', 'breadth', 'of', 'the', 'land', 'in', 'search', 'of', 'knights', 'who', 'will', 'join', 'me', 'in', 'my', 'court', 'at', 'camelot', '.', 'i', 'must', 'speak', 'with', 'your', 'lord', 'and', 'master', '.', 'soldier', '#', '1', ':', 'what', '?', 'ridden', 'on', 'a', 'horse', '?', 'arthur', ':', 'yes', '!', 'soldie

In [80]:
i = inflect.engine()
new_token = []

for t in lowercase_text_token:
  if t.isdigit():
    new_token.append(i.number_to_words(t))
  else:
      new_token.append(t)

print(new_token)

['scene', 'one', ':', '[', 'wind', ']', '[', 'clop', 'clop', 'clop', ']', 'king', 'arthur', ':', 'whoa', 'there', '!', '[', 'clop', 'clop', 'clop', ']', 'soldier', '#', 'one', ':', 'halt', '!', 'who', 'goes', 'there', '?', 'arthur', ':', 'it', 'is', 'i', ',', 'arthur', ',', 'son', 'of', 'uther', 'pendragon', ',', 'from', 'the', 'castle', 'of', 'camelot', '.', 'king', 'of', 'the', 'britons', ',', 'defeator', 'of', 'the', 'saxons', ',', 'sovereign', 'of', 'all', 'england', '!', 'soldier', '#', 'one', ':', 'pull', 'the', 'other', 'one', '!', 'arthur', ':', 'i', 'am', ',', '...', 'and', 'this', 'is', 'my', 'trusty', 'servant', 'patsy', '.', 'we', 'have', 'ridden', 'the', 'length', 'and', 'breadth', 'of', 'the', 'land', 'in', 'search', 'of', 'knights', 'who', 'will', 'join', 'me', 'in', 'my', 'court', 'at', 'camelot', '.', 'i', 'must', 'speak', 'with', 'your', 'lord', 'and', 'master', '.', 'soldier', '#', 'one', ':', 'what', '?', 'ridden', 'on', 'a', 'horse', '?', 'arthur', ':', 'yes', '!',

In [84]:
#token correction
new_txt = " ".join(new_token)
new_txt = new_txt.replace("-"," ")
#print(new_txt)

#words to tokens
unq_token = word_tokenize(new_txt)

#printing 
print("Unique tokens before lowering the case: ", beforelowering_text_token)
print("Number of Unique tokens before lowering the case =", len(beforelowering_text_token))
print("\nUnique tokens after lowering the case: ", unq_token)
print("Number of Unique tokens after lowering the case =", len(unq_token))

Unique tokens before lowering the case:  ['SCENE', '1', ':', '[', 'wind', ']', '[', 'clop', 'clop', 'clop', ']', 'KING', 'ARTHUR', ':', 'Whoa', 'there', '!', '[', 'clop', 'clop', 'clop', ']', 'SOLDIER', '#', '1', ':', 'Halt', '!', 'Who', 'goes', 'there', '?', 'ARTHUR', ':', 'It', 'is', 'I', ',', 'Arthur', ',', 'son', 'of', 'Uther', 'Pendragon', ',', 'from', 'the', 'castle', 'of', 'Camelot', '.', 'King', 'of', 'the', 'Britons', ',', 'defeator', 'of', 'the', 'Saxons', ',', 'sovereign', 'of', 'all', 'England', '!', 'SOLDIER', '#', '1', ':', 'Pull', 'the', 'other', 'one', '!', 'ARTHUR', ':', 'I', 'am', ',', '...', 'and', 'this', 'is', 'my', 'trusty', 'servant', 'Patsy', '.', 'We', 'have', 'ridden', 'the', 'length', 'and', 'breadth', 'of', 'the', 'land', 'in', 'search', 'of', 'knights', 'who', 'will', 'join', 'me', 'in', 'my', 'court', 'at', 'Camelot', '.', 'I', 'must', 'speak', 'with', 'your', 'lord', 'and', 'master', '.', 'SOLDIER', '#', '1', ':', 'What', '?', 'Ridden', 'on', 'a', 'horse'

Answer 2-

Number of Unique tokens before lowering the case = 16450

Number of Unique tokens after lowering the case = 16390


3. Report the number of stopwords in the file. Report the number of tokens left after stopword removal.

In [93]:
#removing the stop words
#
stop_wd = set(stopwords.words('english'))
print("Stopwords in English= ", stop_wd)
print("Total Number of Stopwords in English = ", len(stop_wd))


#to find the stopwords in the given file
#
new_stop_wd = set()  #assigning set
#
for st in stop_wd:
    new_stop_wd.add(st.translate(mapping))  
#  
print("\nStopwords in the sample text- ", new_stop_wd)


#finding new tokens without stopwords
imp_token = [t for t in original_text_token if t not in new_stop_wd]

#printing
print("Tokens after removing Stopwords are- ", imp_token)
print("\nOriginal number of tokens: ",len(original_text_token))
print("Number of tokens after removing Stopwords =", len(imp_token))
print("Number of Stopwords removed from the sample text: ",len(original_text_token)-len(imp_token), " or ",round((len(original_text_token)-len(imp_token))*100/len(original_text_token),2),"%")

Stopwords in English=  {'no', 'we', "she's", "isn't", 'off', 'own', 'into', 'your', 'again', 'hasn', 'their', 'for', 'them', 'with', 'by', 'am', 'of', 'over', 'couldn', 'myself', 'most', "don't", 'shouldn', "weren't", 'once', 'ain', 'any', 'same', 'why', 'under', 'after', 'as', 'should', 'now', 'ourselves', 'more', 'the', 'yourselves', 'in', "didn't", "you'll", "mustn't", 'did', 'a', 'here', 'wasn', 'to', 'very', 'be', 'few', 'her', 'hadn', 'these', 'before', 'further', 'while', 'there', 'mightn', 've', 'below', 'it', "you're", 'have', 'through', 'between', 'too', 'having', 'not', 'shan', 'what', "wouldn't", 'been', 'herself', 'our', 'out', 'he', 'who', 'just', 'i', 'each', 'doing', "you'd", 'being', 'are', 'm', 'about', "shan't", 'nor', 'me', 's', 'his', 'theirs', 'does', "wasn't", 'wouldn', 'at', 'only', 'isn', 'him', "you've", 'which', 'were', 'do', 'until', "hasn't", "won't", 'has', "needn't", 'those', 'they', 'themselves', 'but', 'yourself', 'other', 'if', 'above', 'll', 't', 'you

Answer 3-

Number of tokens after removing Stopwords = 13142


4. Perform stemming after removing stopwords and report the number of unique tokens left in the text.

In [105]:
#Normalising the tokens- Stemming 

unique_tokens = list(set(imp_token)) ##imp_token ---> after removing stopwords (done previously)
unique_tokens.sort()
print("Unique tokens after removing stop words are- ", unique_tokens)
print("Number of Unique/Non-repetitive tokens after removing stop words =", len(unique_tokens))

#Stemming
pr = PorterStemmer()
stemmed_tokens = [pr.stem(t) for t in imp_token]

stemmedunique_tokens = list(set(stemmed_tokens))  #list of stemmed tokens
stemmedunique_tokens.sort()

#printing
print("\nStemmed Tokens are-", stemmedunique_tokens)
print("Number of Stemmed tokens =", len(stemmedunique_tokens))
print("\nNumber of tokens after stopword removal & stemming =", len(imp_token)-len(stemmedunique_tokens))

Unique tokens after removing stop words are-  ['!', '#', "'", "'Aauuuuugh", "'Course", "'Dennis", "'Ecky-ecky-ecky-ecky-pikang-zoop-boing-goodem-zoo-owli-zhiv", "'Erbert", "'Ere", "'First", "'Here", "'I", "'Man", "'Morning", "'Ni", "'O", "'Oooooooh", "'S", "'T", "'Til", "'To", "'aaaah", "'aaggggh", "'anging", "'cause", "'d", "'e", "'em", "'forgive", "'is", "'it", "'ll", "'m", "'ni", "'old", "'re", "'round", "'s", "'shrubberies", "'sorry", "'til", "'uuggggggh", "'ve", '(', ')', ',', '--', '.', '...', '1', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '2', '20', '21', '22', '23', '24', '3', '4', '5', '6', '7', '8', '9', ':', ';', '?', 'A', 'ALL', 'AMAZING', 'ANIMATOR', 'ARMY', 'ARTHUR', 'Aaaaaaaaah', 'Aaaaaaaah', 'Aaaaaah', 'Aaaah', 'Aaaaugh', 'Aaagh', 'Aaah', 'Aaauggh', 'Aaaugh', 'Aaauugh', 'Aagh', 'Aah', 'Aauuggghhh', 'Aauuugh', 'Aauuuves', 'Action', 'Actually', 'African', 'Ages', 'Aggh', 'Agh', 'Ah', 'Ahh', 'Alice', 'All', 'Allo', 'Almighty', 'Alright', 'Am', 'Amen', 'An

Answer4-

Number of tokens after stopword removal & stemming = 11509


5. Report the number of words starting with a consonant and the number of words starting with a vowel in the file given after performing steps 1,2,3, and 4.

In [125]:
#to get back to sample/original text

trace_text = imp_token + list(new_stop_wd)
print(trace_text)
len(trace_text)
#len(imp_token)
#len(original_text_token)-len(imp_token)

['SCENE', '1', ':', '[', 'wind', ']', '[', 'clop', 'clop', 'clop', ']', 'KING', 'ARTHUR', ':', 'Whoa', '!', '[', 'clop', 'clop', 'clop', ']', 'SOLDIER', '#', '1', ':', 'Halt', '!', 'Who', 'goes', '?', 'ARTHUR', ':', 'It', 'I', ',', 'Arthur', ',', 'son', 'Uther', 'Pendragon', ',', 'castle', 'Camelot', '.', 'King', 'Britons', ',', 'defeator', 'Saxons', ',', 'sovereign', 'England', '!', 'SOLDIER', '#', '1', ':', 'Pull', 'one', '!', 'ARTHUR', ':', 'I', ',', '...', 'trusty', 'servant', 'Patsy', '.', 'We', 'ridden', 'length', 'breadth', 'land', 'search', 'knights', 'join', 'court', 'Camelot', '.', 'I', 'must', 'speak', 'lord', 'master', '.', 'SOLDIER', '#', '1', ':', 'What', '?', 'Ridden', 'horse', '?', 'ARTHUR', ':', 'Yes', '!', 'SOLDIER', '#', '1', ':', 'You', "'re", 'using', 'coconuts', '!', 'ARTHUR', ':', 'What', '?', 'SOLDIER', '#', '1', ':', 'You', "'ve", 'got', 'two', 'empty', 'halves', 'coconut', "'re", 'bangin', "'", "'em", 'together', '.', 'ARTHUR', ':', 'So', '?', 'We', 'ridden', 

13320

In [5]:
text= original_text
splitstring= text.split()

v= 0
c= 0

for word in splitstring:
    if word[0] in ['a','e','i','o','u','A','E','I','O','U','[','!','@','#','$','%','^','&','*','(',')']:
        v =v+1
    else: 
        c = c+1

print("Number of words starting with Vowels =",v)
print("Number of words starting with Consonants =", c)

Number of words starting with Vowels = 3216
Number of words starting with Consonants = 7997


Answer 5-

Number of words starting with Vowels = 3216

Number of words starting with Consonants = 7997

6. Given a word and a file as input, return the number of sentences starting with that word in the input file after performing steps 1,2,3, and 4.

In [49]:
#let input word
let_start_word = 'Can'

s=0
if re.findall(r"\.\s*(\w+)", original_text) == let_start_word:
  s =s +1

print("Number of sentences starting with let_start_word =", s)

Number of sentences starting with let_start_word = 0


7. Given a word and a file as input, return the number of sentences ending with that word in the input file after performing steps 1,2,3, and 4.

In [44]:
#let an input word
let_end_word = "them"

text = original_text
end=0

uniq_end_words = set()

for word in text.split():
    if '.' in word:
        # check if period (.) is at the end
        if '.' == word[len(word) -1]:
            uniq_end_words.add(word.rstrip('.'))
l = list(uniq_end_words)
print("list of all end words in a given sentence of the text", l)


for i in range(len(l)):
  if l[i] == let_word:
    end =end +1
print("\nNumber of sentences ending with the givenn let_end_word =", end)

list of all end words in a given sentence of the text ['', 'one', 'Launcelot', 'alarm', 'scene', 'rabbit', 'there', 'lair', 'Brother', 'logically', 'together', 'Ah', 'Well', 'man', 'were', 'thirty-seven', 'Gable', "'", 'Ohh', 'swallow', 'Britons', 'k-nnnnniggets', 'fled', "didn't", 'Grail', 'glad', 'guard', 'country', 'able', 'Nothing', 'shit', 'would', 'mistake', 'model', 'Quite', 'watch', 'Yup', 'go', 'big', 'peril', 'van', 'person', 'guest', 'Follow', 'blanket', 'enchanter', 'Galahad', 'yours', 'same', 'along', 'less', 'single-handed', 'clear', 'Say', 'varletesses', 'find', 'Hmm', 'earthquakes', 'me', 'commune', 'ladies', 'simple', 'this', 'Nay', 'very', 'lads', '[', 'bridge', 'get', "don't", 'cope', 'can', 'islands', 'Winter', 'ho', 'Blue', 'leave', 'in', 'ones', 'heh', 'attack', 'sir', 'say', 'assault', 'table', 'explain', 'chance', 'zone', "'uuggggggh'", 'woods', 'system', 'knights', 'see', 'scholar', 'now', 'today', 'Ninepence', 'worse', 'ninepence', 'Umhm', 'then', 'rejoicing',

8. Given a word and a file as input, return the count of that word in the input file after performing steps 1,2,3, and 4.

In [57]:
#let an input
let_word = "Who"

text = original_text
a=[]
count=0
a=text.split(" ")
#print(a)

for i in range(0,len(a)):
      if(let_word==a[i]):
            count=count+1
print("Count of the given word in the text =", count)


Count of the given word in the text = 21
